agent/ebs/watcher.go

```diff
@@ -0,0 +1,212 @@
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package ebs
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/aws/amazon-ecs-agent/agent/engine/dockerstate"
+	"github.com/aws/amazon-ecs-agent/agent/statechange"
+	apiebs "github.com/aws/amazon-ecs-agent/ecs-agent/api/resource"
+	log "github.com/cihub/seelog"
+)
+
+type EBSWatcher struct {
+	ctx        context.Context
+	cancel     context.CancelFunc
+	scanTicker *time.Ticker
+	agentState dockerstate.TaskEngineState
+	// TODO: The dataClient will be used to save to agent's data client as well as start the ACK timer. This will be added once the data client functionality have been added
+	// dataClient     data.Client
+	ebsChangeEvent  chan<- statechange.Event
+	discoveryClient apiebs.EBSDiscovery
+	mailbox         chan func()
+}
+
+// NewWatcher is used to return a new instance of the EBSWatcher struct
+func NewWatcher(ctx context.Context,
+	state dockerstate.TaskEngineState,
+	stateChangeEvents chan<- statechange.Event) (*EBSWatcher, error) {
+	derivedContext, cancel := context.WithCancel(ctx)
+	discoveryClient := apiebs.NewDiscoveryClient(derivedContext)
+	return &EBSWatcher{
+		ctx:             derivedContext,
+		cancel:          cancel,
+		agentState:      state,
+		ebsChangeEvent:  stateChangeEvents,
+		discoveryClient: discoveryClient,
+		mailbox:         make(chan func(), 100),
+	}, nil
+}
+
+// Start is used to kick off the periodic scanning process of the EBS volume attachments for the EBS watcher.
+// If there aren't any initially, the scan ticker will stop.
+func (w *EBSWatcher) Start() {
+	log.Info("Starting EBS watcher.")
+
+	w.scanTicker = time.NewTicker(apiebs.ScanPeriod)
+	if len(w.agentState.GetAllPendingEBSAttachments()) == 0 {
+		w.scanTicker.Stop()
+	}
+
+	for {
+		select {
+		case f := <-w.mailbox:
+			f()
+		case <-w.scanTicker.C:
+			pendingEBS := w.agentState.GetAllPendingEBSAttachmentsWithKey()
+			foundVolumes := apiebs.ScanEBSVolumes(pendingEBS, w.discoveryClient)
+			w.NotifyFound(foundVolumes)
+		case <-w.ctx.Done():
+			w.scanTicker.Stop()
+			log.Info("EBS Watcher Stopped")
+			return
+		}
+	}
+}
+
+// Stop will stop the EBS watcher
+func (w *EBSWatcher) Stop() {
+	log.Info("Stopping EBS watcher.")
+	w.cancel()
+}
+
+// HandleResourceAttachment processes the resource attachment message.
+func (w *EBSWatcher) HandleResourceAttachment(ebs *apiebs.ResourceAttachment) error {
+	var err error
+	var wg sync.WaitGroup
+	wg.Add(1)
+	w.mailbox <- func() {
+		defer wg.Done()
+		empty := len(w.agentState.GetAllPendingEBSAttachments()) == 0
+
+		err := w.handleEBSAttachment(ebs)
+		if err != nil {
+			log.Warnf("Failed to handle resource attachment %v", ebs.String())
+		}
+		if empty && len(w.agentState.GetAllPendingEBSAttachments()) == 1 {
```

Comment:
Can we please add some comments explaining this condition. It's not very clear to me.

Reply:
Yep, renamed both `empty` and added some comments to better explain.